genie-core/src/main/java/com/netflix/genie/core/services/JobCoordinatorService.java

```diff
@@ -136,30 +136,40 @@ public String coordinateJob(
             .withId(jobRequest.getId())
             .withTags(jobRequest.getTags());
 
-        if (this.canRunJob()) {
-            jobBuilder
-                .withStatus(JobStatus.INIT)
-                .withStatusMsg("Job Accepted and in initialization phase.");
-            // TODO: if this throws exception the job will never be marked failed
-            this.jobPersistenceService.createJob(jobBuilder.build());
-            try {
-                final Future<?> task
-                    = this.taskExecutor.submit(new JobLauncher(this.jobSubmitterService, jobRequest, this.registry));
+        synchronized (this) {
```

Comment:
This fix looks correct to really coordinate the max number of jobs on an instance. Instead of synchronized and getting the count, cant we have a thread pool with a queue capacity of max jobs. If the number goes up, the pool will reject it and you can either return back with a failure or redirect to another instance.